### PR TITLE
Storage Blob Stg78 Features

### PR DESCRIPTION
**Pull request by Jinming-Hu**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp [hunk 1]
```diff
@@ -402,6 +402,10 @@ namespace Azure { namespace Storage { namespace Blobs {
        * container was deleted.
        */
       Nullable<int32_t> RemainingRetentionDays;
+      /**
+       * Indicates if version level worm is enabled on this container.
+       */
+      bool HasImmutableStorageWithVersioning = false;
```
Comment (katmsft) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp:408]: Any change log needed for this?

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp [hunk 2]
```diff
@@ -2200,6 +2207,218 @@ namespace Azure { namespace Storage { namespace Blobs {
...
+         */
+        std::string Expression;
+        Nullable<QuerySerialization> InputSerialization;
+        Nullable<QuerySerialization> OutputSerialization;
+      };
+    } // namespace _detail
+    /**
+     * @brief Response type for #Azure::Storage::Blobs::BlobClient::Query.
+     */
+    struct QueryBlobResult final
```
Comment (katmsft) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp:2388]: Seems to be missing some fields:
Content-Encoding
Content-Language
Cache-Control
Content-Disposition
Access-Control-Allow-Origin
Access-Control-Expose-Headers
Vary
Access-Control-Allow-Credentials
x-ms-blob-committed-block-count
x-ms-server-encrypted

File: sdk/storage/azure-storage-blobs/CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,12 @@
 
 ### Features Added
 
+- Bumped up API version to `2020-10-02`.
+- Added new API: `BlockBlobClient::Query()`.
+- Added `ContinuationToken` and `PageSizeHint` in `GetPageRangesOptions`.
+- Added support for listing system containers.
+- Added support for listing deleted root blob with active versions.
+
```
Comment (katmsft) [inline sdk/storage/azure-storage-blobs/CHANGELOG.md:12]: Is source auth with OAuth for Copy Blob API supported? I only see the change in the protocol layer.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp (hunk 2 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp:2388]: Where did you get these?
In .Net, they just reuse the struct for `Download()`. But apparently that's wrong.
I've tested, storage service doesn't return all those headers as stated in swagger file/API reference.

So the situation is, we don't know which headers will be returned and which won't. I don't think customers will expect to get those headers from `Query()` API anyway.

File: sdk/storage/azure-storage-blobs/test/ut/blob_query_test.cpp [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+    {
+      queryOptions.OutputTextConfiguration
+          = Blobs::BlobQueryOutputTextOptions::CreateCsvTextOptions();
+      auto queryResponse
+          = client.Query("SELECT * from BlobStorage WHERE id > 101 AND price < 100;", queryOptions);
+      auto data = queryResponse.Value.BodyStream->ReadToEnd();
+      EXPECT_EQ(
+          std::string(data.begin(), data.end()),
+          R"csv(103,apples,99
+106,lemons,69
```
Comment (katmsft) [inline sdk/storage/azure-storage-blobs/test/ut/blob_query_test.cpp:92]: nic: indent.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/test/ut/blob_query_test.cpp:92]: This is not indentation issue. It's called raw string literal.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp (hunk 2 above)
Comment (katmsft) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp:2388]: Got it from the official doc:
https://docs.microsoft.com/en-us/rest/api/storageservices/query-blob-contents
Might be a doc issue, please double check.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp [hunk 5]
```diff
@@ -2200,6 +2207,218 @@ namespace Azure { namespace Storage { namespace Blobs {
...
+      {
+        /**
+         * Array of BlobQueryArrowField.
+         */
+        std::vector<BlobQueryArrowField> Schema;
+      };
+      /**
+       * @brief Parquet configuration.
+       */
+      struct ParquetConfiguration final
```
[the remark this answers is not in the document]
Comment (katmsft) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp:2336]: I mean, are there more configurable fields for `ParquetConfiguration` other than null/not null? If not, should we just expose a boolean in the option?
Is this preserved for forward compatibility?

File: sdk/storage/azure-storage-blobs/test/ut/blob_query_test.cpp (hunk 4 above)
Comment (katmsft) [reply sdk/storage/azure-storage-blobs/test/ut/blob_query_test.cpp:92]: line 49 to line 69 in this file is handling raw string literal well.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp (hunk 2 above)
Comment (katmsft) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp:2388]: According to the doc, most of the headers listed above only return under a certain condition. I am fine with you adding it later but from a feature completeness perspective, it would be nice to have them in version 1.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (katmsft) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp:2336]: > In the format field we must mention the string parquet and the ParquetTextConfiguration doesn’t have any child element in it

Found this in the design doc. I must say that changing to a boolean is much less confusing here.

File: sdk/storage/azure-storage-blobs/src/private/avro_parser.cpp [hunk 6]
```diff
@@ -0,0 +1,680 @@
...
+    {
+      uint64_t r = 0;
+      int nb = 0;
+      while (true)
+      {
+        uint8_t c = (*data.BufferPtr)[data.Offset++];
+        r = r | ((static_cast<uint64_t>(c) & 0x7f) << (nb * 7));
+        if (c & 0x80)
+        {
+          ++nb;
```
Comment (katmsft) [inline sdk/storage/azure-storage-blobs/src/private/avro_parser.cpp:25]: I am confused as why we are moving 7 digits per iteration, e.g. for -128, it has only 1 byte, however, it is reading to second iteration, which reads another byte from the buffer.

could you add some test cases for negative numbers?

File: sdk/storage/azure-storage-blobs/test/ut/blob_query_test.cpp (hunk 4 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/test/ut/blob_query_test.cpp:92]: The grammar is different.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp:2336]: Generator doesn't work that way. If it was a boolean, the generated xml would be:

```xml
<ParquetTextConfiguration>true</ParquetTextConfiguration>
```

File: sdk/storage/azure-storage-blobs/src/private/avro_parser.cpp (hunk 6 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/src/private/avro_parser.cpp:25]: Look at the binary encoding section at this doc https://avro.apache.org/docs/current/spec.html

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/src/private/avro_parser.cpp:25]: > could you add some test cases for negative numbers?

We don't test internal functions.

Integer parsing is a very basic functionality. If the whole avro parser works, I assume this function is good.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp (hunk 2 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp:2388]: I decided to hold it off until we have customer request.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (katmsft) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp:2336]: I'm not referring to the protocol layer code. In convenience layer options we can hide this detail.

File: sdk/storage/azure-storage-blobs/src/private/avro_parser.cpp (hunk 6 above)
Comment (katmsft) [reply sdk/storage/azure-storage-blobs/src/private/avro_parser.cpp:25]: My concern is that this is actually tested for positive numbers but not negative numbers. If you have confidence that's fine, but wouldn't hurt for some extra coverage.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp:2336]: [Here](https://github.com/Azure/azure-sdk-for-cpp/blob/a798e02887fc200c3f66c847f1b5c93fd03a921d/sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp#L1079)'s convenience layer code. Does it look good to you?

**Review by katmsft — APPROVED** [review]
Approved with some suggestions.
